### DISCUSSION TITLE
Centreline Extraction

### DISCUSSION THREAD

**Discussion by swtb3** [pyvista/pyvista#4504] [Q&A] [answered]
Hi, 

I'm working with vascular structures and it would be useful to extract a centreline from the model. Is there a way to do this in pyvista?

Cheers!

**Comment by banesullivan** [comment]
Would you please clarify what you mean by "centerline"? Is this a line through the center of the extents/bounds? Does it go from top to bottom? lef-right? along the diagonal?

Please clarify and perhaps annotate a screenshot to help us understand what you are after

**Reply by swtb3** [reply]
So the model would be a tubular structure. The centre line is a line that starts at one end of the tube moving towards the other end keeping equidistant from the surface containing it.

**Reply by banesullivan** [reply]
Ah I see now. We do not have a way to do this out of the box with PyVista as that's a rather complex algorithm. As @beroda points out below, vmtk is probably the toolkit for that sort of algorithm

**Reply by swtb3** [reply]
I have found the vmtk extension to be quite diffcult to use as they seem to be based around the 3d slicer. I do have an algorithm that will find the centreline from a numpy array. However I am stuggling to convert from the uniformgrid produced from the Nifti reader to a numpy array. it seems to distort my model a fair bit. Do you have any guidance on that conversion?

**Reply by banesullivan** [reply]
> However I am stuggling to convert from the uniformgrid produced from the Nifti reader to a numpy array. it seems to distort my model a fair bit.

What do you mean by distort? It shouldn't alter the data in any way.

You can get a 3D numpy array from the UniformGrid by:

```py
ndarray = grid['name'].reshape(grid.dimensions)
```

**Reply by swtb3** [reply]
Hi, Thanks for getting back to me.

When I use the code above to extract the points as an array the plot is looking like this:[](url)
![Capture](https://github.com/pyvista/pyvista/assets/135991636/4739e57a-6c85-426e-b6e9-e5fba6977c7e)

For context, the plot should show a segment of the Aorta.

my version of the code looks like :

`ndarray = mesh['NIFTI'].reshape(mesh.dimensions)`

before this I do also change the spacing to be (1, 1, 1)

if i do `mesh.plot()` this show the correct plot.

**Reply by banesullivan** [reply]
Can you share the data file?

**Reply by swtb3** [reply]
Sure thing, thank you for the assistance :)

[aorta.nii.gz](https://github.com/pyvista/pyvista/files/11713214/aorta.nii.gz)

**Reply by banesullivan** [reply]
Try `order='F'` when reshaping, my bad: `ndarray = mesh['NIFTI'].reshape(mesh.dimensions, order='F')`

**Reply by swtb3** [reply]
That sorted it! thank you ! Marking beroda as the answer since it is techically the answer to the origional query :)

**Comment by edabor** [comment, marked as answer]
Hi,

To work with vascular structures, **vmtk** is the tool you need.
https://github.com/vmtk
http://www.vmtk.org/index.html

It is built from **VTK**, thus geometries produced with **vmtk** could be easily read with **pyvista**.

However, the **vmtk** project seems to be unmaintained, but a plugin for 3D slicer has been developed.
https://github.com/vmtk/SlicerExtension-VMTK and 3D slicer is very active. You may post questions in their discourse forum there: https://discourse.slicer.org/
